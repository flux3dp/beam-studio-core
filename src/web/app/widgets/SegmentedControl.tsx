import * as React from 'react';
import classNames from 'classnames';

interface Props {
  segments: {
    imgSrc: string;
    title: string;
    value: number;
  }[];
  selectedIndexes: number[];
  isDisabled: boolean;
  onChanged: (type: number) => void;
}

function SegmentedControl({
  segments, selectedIndexes, isDisabled, onChanged,
}: Props): JSX.Element {
  if (selectedIndexes.length > 1) {
    console.warn('Selected more than 1 items for exclusive control');
    selectedIndexes.splice(1, selectedIndexes.length - 1);
  }

  const onClick = (index) => {
    if (isDisabled) {
      return;
    }
    if (selectedIndexes[0] !== index) {
      const value = segments[index].value || index;
      onChanged(value);
    }
  };

  const renderItems = () => {
    const items = [];
    for (let i = 0; i < segments.length; i += 1) {
      const segment = segments[i];
      const isSelected = selectedIndexes.includes(i);
      items.push(
        <div key={i} className={classNames('seg-item', { selected: isSelected })} title={segment.title} onClick={() => onClick(i)}>
          <img src={segment.imgSrc} className="seg-item-image" />
        </div>,
      );
    }
    return items;
  };

  return (
    <div className={classNames('segmented-control', { disabled: isDisabled })}>
      {renderItems()}
    </div>
  );
}

export default SegmentedControl;
