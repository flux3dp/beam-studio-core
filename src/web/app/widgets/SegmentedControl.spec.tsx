/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import SegmentedControl from './SegmentedControl';

describe('test SegmentedControl', () => {
  test('enabled', () => {
    const onChanged = jest.fn();
    const wrapper = shallow(<SegmentedControl
      isDisabled={false}
      selectedIndexes={[0]}
      onChanged={onChanged}
      segments={[
        {
          imgSrc: 'img/right-panel/icon-nodetype-0.svg',
          title: 'tCorner',
          value: 0,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-1.svg',
          title: 'tSmooth',
          value: 1,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-2.svg',
          title: 'tSymmetry',
          value: 2,
        },
      ]}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('div[title="tSmooth"]').simulate('click');
    expect(onChanged).toHaveBeenCalledTimes(1);
    expect(onChanged).toHaveBeenNthCalledWith(1, 1);
  });

  test('disabled', () => {
    const onChanged = jest.fn();
    const wrapper = shallow(<SegmentedControl
      isDisabled
      selectedIndexes={[0]}
      onChanged={onChanged}
      segments={[
        {
          imgSrc: 'img/right-panel/icon-nodetype-0.svg',
          title: 'tCorner',
          value: 0,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-1.svg',
          title: 'tSmooth',
          value: 1,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-2.svg',
          title: 'tSymmetry',
          value: 2,
        },
      ]}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('div[title="tSmooth"]').simulate('click');
    expect(onChanged).not.toHaveBeenCalled();
  });

  test('click on selected index', () => {
    const onChanged = jest.fn();
    const wrapper = shallow(<SegmentedControl
      isDisabled={false}
      selectedIndexes={[0]}
      onChanged={onChanged}
      segments={[
        {
          imgSrc: 'img/right-panel/icon-nodetype-0.svg',
          title: 'tCorner',
          value: 0,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-1.svg',
          title: 'tSmooth',
          value: 1,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-2.svg',
          title: 'tSymmetry',
          value: 2,
        },
      ]}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find('div[title="tCorner"]').simulate('click');
    expect(onChanged).not.toHaveBeenCalled();
  });

  test('multiple selected indexes', () => {
    const onChanged = jest.fn();
    const wrapper = shallow(<SegmentedControl
      isDisabled={false}
      selectedIndexes={[0, 1, 2]}
      onChanged={onChanged}
      segments={[
        {
          imgSrc: 'img/right-panel/icon-nodetype-0.svg',
          title: 'tCorner',
          value: 0,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-1.svg',
          title: 'tSmooth',
          value: 1,
        },
        {
          imgSrc: 'img/right-panel/icon-nodetype-2.svg',
          title: 'tSymmetry',
          value: 2,
        },
      ]}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
