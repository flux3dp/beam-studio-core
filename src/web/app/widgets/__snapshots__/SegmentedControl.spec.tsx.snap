// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SegmentedControl click on selected index 1`] = `
<div
  className="segmented-control"
>
  <div
    className="seg-item selected"
    key="0"
    onClick={[Function]}
    title="tCorner"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-0.svg"
    />
  </div>
  <div
    className="seg-item"
    key="1"
    onClick={[Function]}
    title="tSmooth"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-1.svg"
    />
  </div>
  <div
    className="seg-item"
    key="2"
    onClick={[Function]}
    title="tSymmetry"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-2.svg"
    />
  </div>
</div>
`;

exports[`test SegmentedControl disabled 1`] = `
<div
  className="segmented-control disabled"
>
  <div
    className="seg-item selected"
    key="0"
    onClick={[Function]}
    title="tCorner"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-0.svg"
    />
  </div>
  <div
    className="seg-item"
    key="1"
    onClick={[Function]}
    title="tSmooth"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-1.svg"
    />
  </div>
  <div
    className="seg-item"
    key="2"
    onClick={[Function]}
    title="tSymmetry"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-2.svg"
    />
  </div>
</div>
`;

exports[`test SegmentedControl enabled 1`] = `
<div
  className="segmented-control"
>
  <div
    className="seg-item selected"
    key="0"
    onClick={[Function]}
    title="tCorner"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-0.svg"
    />
  </div>
  <div
    className="seg-item"
    key="1"
    onClick={[Function]}
    title="tSmooth"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-1.svg"
    />
  </div>
  <div
    className="seg-item"
    key="2"
    onClick={[Function]}
    title="tSymmetry"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-2.svg"
    />
  </div>
</div>
`;

exports[`test SegmentedControl multiple selected indexes 1`] = `
<div
  className="segmented-control"
>
  <div
    className="seg-item selected"
    key="0"
    onClick={[Function]}
    title="tCorner"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-0.svg"
    />
  </div>
  <div
    className="seg-item"
    key="1"
    onClick={[Function]}
    title="tSmooth"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-1.svg"
    />
  </div>
  <div
    className="seg-item"
    key="2"
    onClick={[Function]}
    title="tSymmetry"
  >
    <img
      className="seg-item-image"
      src="img/right-panel/icon-nodetype-2.svg"
    />
  </div>
</div>
`;
